File: fannkuch-redux/Version6/ChatGPT/funnkuch-redux6.c
// fannkuch-redux.c — C port of fannkuch-redux6.py (Computer Language Benchmarks Game)
// Build:  gcc -O3 -march=native -o fannkuch fannkuch-redux.c
// Run:    ./fannkuch 12
//
// Prints:
//   <checksum>
//   Pfannkuchen(<n>) = <maxFlips>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static inline void reverse_prefix(int *a, int k) {
    int i = 0, j = k;
    while (i < j) { int t = a[i]; a[i] = a[j]; a[j] = t; ++i; --j; }
}

static int fannkuch(int n, long long *checksum_out) {
    int *perm1 = (int*)malloc(n * sizeof(int));
    int *count = (int*)malloc(n * sizeof(int));
    int *perm  = (int*)malloc(n * sizeof(int));
    if (!perm1 || !count || !perm) { fprintf(stderr, "oom\n"); exit(1); }

    for (int i = 0; i < n; ++i) { perm1[i] = i; count[i] = i; }

    int maxFlipsCount = 0;
    int permSign = 1;                 // True (1) / False (0) in Python
    long long checksum = 0;
    const int nm = n - 1;

    for (;;) {
        int k = perm1[0];
        if (k != 0) {
            memcpy(perm, perm1, n * sizeof(int));
            int flipsCount = 1;
            int kk = perm[k];
            while (kk != 0) {
                reverse_prefix(perm, k);
                ++flipsCount;
                k  = kk;
                kk = perm[kk];
            }
            if (maxFlipsCount < flipsCount) maxFlipsCount = flipsCount;
            checksum += (permSign ? flipsCount : -flipsCount);
        }

        // Generate next permutation (incremental change)
        if (permSign) {
            int t = perm1[0]; perm1[0] = perm1[1]; perm1[1] = t;
            permSign = 0; // False
        } else {
            int t = perm1[1]; perm1[1] = perm1[2]; perm1[2] = t;
            permSign = 1; // True

            int r;
            int broke = 0;
            for (r = 2; r < n - 1; ++r) {
                if (count[r] != 0) { broke = 1; break; }
                count[r] = r;
                int perm0 = perm1[0];
                // rotate left perm1[0..r] by 1: move perm1[1..r] down, put perm0 at r
                memmove(&perm1[0], &perm1[1], r * sizeof(int));
                perm1[r] = perm0;
            }
            if (!broke) {
                r = nm;
                if (count[r] == 0) {
                    // Done: produce checksum and return max
                    *checksum_out = checksum;
                    free(perm); free(count); free(perm1);
                    return maxFlipsCount;
                }
            }
            count[r] -= 1;
        }
    }
}

int main(int argc, char **argv) {
    if (argc < 2) { fprintf(stderr, "usage: %s <n>\n", argv[0]); return 1; }
    int n = atoi(argv[1]);
    if (n < 1) { fprintf(stderr, "n must be >= 1\n"); return 1; }

    long long checksum = 0;
    int maxFlips = fannkuch(n, &checksum);

    printf("%lld\n", checksum);
    printf("Pfannkuchen(%d) = %d\n", n, maxFlips);
    return 0;
}

File: fannkuch-redux/Version6/ChatGPT/funnkuch-redux6.rs
use std::env;

/// Reverse prefix perm[0..=k]
#[inline]
fn reverse_prefix(perm: &mut [usize], mut k: usize) {
    let mut i = 0usize;
    while i < k {
        perm.swap(i, k);
        i += 1;
        if k == 0 { break; }
        k -= 1;
    }
}

/// Compute fannkuch for n; returns (checksum, max_flips)
fn fannkuch(n: usize) -> (i64, i32) {
    // perm1: current permutation in generator order
    // count: rotation counters for each prefix
    let mut perm1: Vec<usize> = (0..n).collect();
    let mut count: Vec<i32> = (0..n as i32).collect();

    let mut perm: Vec<usize> = vec![0; n];

    let mut perm_sign = true; // True/False like Python
    let mut max_flips = 0i32;
    let mut checksum = 0i64;
    let nm = n - 1;

    loop {
        // Work (flip counting)
        let k0 = perm1[0];
        if k0 != 0 {
            perm.copy_from_slice(&perm1);
            let mut flips = 1i32;
            let mut k = k0;
            let mut kk = perm[k];
            while kk != 0 {
                reverse_prefix(&mut perm, k);
                flips += 1;
                k = kk;
                kk = perm[kk];
            }
            if flips > max_flips {
                max_flips = flips;
            }
            checksum += if perm_sign { flips as i64 } else { -(flips as i64) };
        }

        // Next permutation (incremental change, matches fannkuch-redux6.py)
        if perm_sign {
            perm1.swap(0, 1);
            perm_sign = false;
        } else {
            perm1.swap(1, 2);
            perm_sign = true;

            let mut r = 2usize;
            let mut broke = false;
            while r < n - 1 {
                if count[r] != 0 {
                    broke = true;
                    break;
                }
                count[r] = r as i32;
                // rotate-left perm1[0..=r] by 1
                let p0 = perm1[0];
                perm1.copy_within(1..=r, 0);
                perm1[r] = p0;
                r += 1;
            }
            if !broke {
                r = nm;
                if count[r] == 0 {
                    // Done
                    return (checksum, max_flips);
                }
            }
            count[r] -= 1;
        }
    }
}

fn main() {
    let n: usize = env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(7);

    let (checksum, max_flips) = fannkuch(n);
    println!("{}", checksum);
    println!("Pfannkuchen({}) = {}", n, max_flips);
}
